pageserver/src/tenant/storage_layer/inmemory_layer.rs

```diff
@@ -326,24 +304,210 @@ impl InMemoryLayer {
+        struct MergedRead<'a> {
+            start_chunk_no: u32,
+            nchunks: u32,
+            dsts: Vec<MergedChunkReadDestination<'a>>,
+        }
+        struct MergedChunkReadDestination<'a> {
+            value_read: &'a ValueRead,
+            offset_in_merged_read: u32,
+            len: u32,
+        }
+        let mut merged_reads: Vec<MergedRead> = Vec::new();
+        let mut chunk_reads = chunk_reads.into_iter().peekable();
+        loop {
+            let mut last_chunk_no = None;
+            let to_merge: Vec<(u32, Vec<ChunkReadDestination>)> = chunk_reads
+                .peeking_take_while(|(chunk_no, _)| {
+                    if let Some(last_chunk_no) = last_chunk_no {
+                        if *chunk_no != last_chunk_no + 1 {
+                            return false;
+                        }
+                    }
+                    last_chunk_no = Some(*chunk_no);
+                    true
+                })
+                .take(MAX_CHUNK_BATCH_SIZE)
+                .collect(); // TODO: avoid this .collect()
+            let Some(start_chunk_no) = to_merge.first().map(|(chunk_no, _)| *chunk_no) else {
+                break;
+            };
+            let nchunks = to_merge.len() as u32;
+            let dsts = to_merge
+                .into_iter()
+                .enumerate()
+                .flat_map(|(i, (_, dsts))| {
+                    dsts.into_iter().map(
+                        move |ChunkReadDestination {
+                                  value_read,
+                                  offset_in_chunk,
+                                  len,
+                              }| {
+                            MergedChunkReadDestination {
+                                value_read,
+                                offset_in_merged_read: i as u32 * DIO_CHUNK_SIZE as u32
+                                    + offset_in_chunk,
+                                len,
+                            }
+                        },
+                    )
+                })
+                .collect();
+            merged_reads.push(MergedRead {
+                start_chunk_no,
+                nchunks,
+                dsts,
+            });
+        }
+        drop(chunk_reads);
+
+        // Execute reads and fill the destination
+        // TODO: prefetch
+        let get_chunk_buf = |nchunks| Vec::with_capacity(nchunks as usize * (DIO_CHUNK_SIZE));
+        for MergedRead {
+            start_chunk_no,
+            nchunks,
+            dsts,
+        } in merged_reads
+        {
+            let all_done = dsts
+                .iter()
+                .all(|MergedChunkReadDestination { value_read, .. }| {
+                    let value_buf = value_read.value_buf.lock().unwrap();
+                    let Ok(buf) = &*value_buf else {
+                        return true; // on Err() there's no need to read more
+                    };
+                    buf.len() == value_read.len as usize
```

Comment:
Hmm, yeah, that's true. Are you implying I should remove the `Result<>` state because the buffer length is sufficient?
IMO we need to `Result<>` so we don't issue useless remaining reads after a failed read.

Reply:
Wasn't suggesting that we remove the result, just pointing out the quirk. Perhaps there's a better way of structuring this. Treat it as a nit tho.